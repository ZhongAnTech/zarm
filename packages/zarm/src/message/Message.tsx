import React, { forwardRef, useState, useContext } from 'react';
import type { HTMLAttributes, MouseEvent } from 'react';
import classnames from 'classnames';
import { ArrowRight as ArrowRightIcon, Close as CloseIcon } from '@zarm-design/icons';
import { ConfigContext } from '../n-config-provider';
import type { BaseMessageProps } from './interface';

export type MessageProps = BaseMessageProps & HTMLAttributes<HTMLDivElement>;

const Message = forwardRef<HTMLDivElement, MessageProps>((props, ref) => {
  const {
    className,
    theme,
    icon,
    hasArrow,
    closable,
    children,
    onClick,
    onClose,
    ...restProps
  } = props;
  const [visible, setVisible] = useState(true);
  const { prefixCls: globalPrefixCls } = useContext(ConfigContext);

  if (!visible) return null;

  const prefixCls = `${globalPrefixCls}-message`;
  const cls = classnames(prefixCls, className, {
    [`${prefixCls}--${theme}`]: theme,
    [`${prefixCls}--link`]: hasArrow,
  });

  function handleClick(e: MouseEvent<HTMLDivElement>) {
    if (!hasArrow) {
      e.preventDefault();
      return;
    }
    onClick?.(e);
  }

  function handleClose(e: MouseEvent<HTMLElement>) {
    setVisible(false);
    onClose?.(e);
  }

  return (
    <div ref={ref} className={cls} onClick={handleClick} {...restProps}>
      {icon && (
        <div className={`${prefixCls}__header`}>
          <div className={`${prefixCls}__icon`}>{icon}</div>
        </div>
      )}
      <div className={`${prefixCls}__body`}>{children}</div>
      {(closable || hasArrow) && (
        <div className={`${prefixCls}__footer`}>
          {hasArrow && <ArrowRightIcon />}
          {closable && <CloseIcon onClick={handleClose} />}
        </div>
      )}
    </div>
  );
});

Message.displayName = 'Message';

Message.defaultProps = {
  theme: 'primary',
  hasArrow: false,
  closable: false,
};

export default Message;
