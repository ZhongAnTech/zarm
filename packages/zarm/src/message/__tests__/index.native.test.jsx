import React from 'react';
import { Alert } from 'react-native';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Message from '../index.native';

describe('Message', () => {
  it('renders correctly', () => {
    const wrapper = render(<Message>This is a Message</Message>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('theme', () => {
    const wrapper = render(<Message theme="danger">This is a Message</Message>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('onClick', () => {
    const wrapper = render(
      <Message onClick={() => Alert.alert('click this message!')}>Message</Message>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('closable', () => {
    const wrapper = render(<Message closable>可关闭</Message>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
