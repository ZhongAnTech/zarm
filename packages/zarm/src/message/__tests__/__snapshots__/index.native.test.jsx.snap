// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message closable 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock>
      可关闭
    </react-native-mock>
  </react-native-mock>
  <react-native-mock>
    <react-native-mock>
      <react-native-mock>
        <react-native-mock />
        <react-native-mock />
      </react-native-mock>
    </react-native-mock>
  </react-native-mock>
</react-native-mock>
`;

exports[`Message onClick 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock>
      <react-native-mock>
        Message
      </react-native-mock>
    </react-native-mock>
  </react-native-mock>
</react-native-mock>
`;

exports[`Message renders correctly 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock>
      This is a Message
    </react-native-mock>
  </react-native-mock>
</react-native-mock>
`;

exports[`Message theme 1`] = `
<react-native-mock>
  <react-native-mock>
    <react-native-mock>
      This is a Message
    </react-native-mock>
  </react-native-mock>
</react-native-mock>
`;
